homeassistant/components/switcher_kis/switch.py

```diff
@@ -0,0 +1,230 @@
+"""Home Assistant Switcher Component Switch platform."""
+
+from datetime import datetime
+from logging import getLogger
+from typing import Callable, cast, Dict, Optional
+
+from homeassistant.components.switch import (
+    ATTR_CURRENT_POWER_W, ENTITY_ID_FORMAT as SWITCH_FORMAT, SwitchDevice)
+from homeassistant.const import CONF_ICON, CONF_NAME
+from homeassistant.core import HomeAssistant, Event
+from homeassistant.exceptions import PlatformNotReady
+
+from . import (
+    ATTR_AUTO_OFF_SET, ATTR_DEVICE_NAME, ATTR_ELECTRIC_CURRNET,
+    ATTR_IP_ADDRESS, ATTR_LAST_DATA_UPDATE, ATTR_LAST_STATE_CHANGE,
+    ATTR_REMAINING_TIME, DISCOVERY_CONFIG, DISCOVERY_DEVICE, DOMAIN,
+    ENTITY_ID_FORMAT as SWITCHER_KIS_FORMAT, EVENT_SWITCHER_DEVICE_UPDATED,
+    UPDATED_DEVICE)
+
+_LOGGER = getLogger(__name__)
+
+DEPENDENCIES = ['switcher_kis']
+
+ENTITY_ID_FORMAT = SWITCH_FORMAT.format(SWITCHER_KIS_FORMAT)
+
+PROPERTIES_TO_ATTRIBUTES = {
+    'current_power_w': ATTR_CURRENT_POWER_W,
+    'device_ip_addr': ATTR_IP_ADDRESS,
+    'electric_current': ATTR_ELECTRIC_CURRNET,
+    'remaining_time': ATTR_REMAINING_TIME,
+    'auto_off_set': ATTR_AUTO_OFF_SET,
+    'last_data_update': ATTR_LAST_DATA_UPDATE,
+    'last_state_change': ATTR_LAST_STATE_CHANGE,
+    'device_name': ATTR_DEVICE_NAME
+}
+
+
+async def async_setup_platform(hass: HomeAssistant, config: Dict,
+                               async_add_entities: Callable,
+                               discovery_info: Dict) -> None:
+    """Set up the switcher platform for the switch component."""
+    if hass.data[DOMAIN][DISCOVERY_CONFIG]:
+        name = hass.data[DOMAIN][DISCOVERY_CONFIG][CONF_NAME].title()
+        icon = hass.data[DOMAIN][DISCOVERY_CONFIG].get(CONF_ICON)
+    else:
+        raise PlatformNotReady("No config data found")
+
+    if hass.data[DOMAIN].get(DISCOVERY_DEVICE):
+        switcher_entity = SwitcherControl(
+            hass, name, icon, hass.data[DOMAIN].get(DISCOVERY_DEVICE))
+    else:
+        raise PlatformNotReady("No device data discoverd")
+
+    async_add_entities([switcher_entity], False)
+
+    return None
+
+
+class SwitcherControl(SwitchDevice):
+    """Home Assistant switch entity."""
+
+    from aioswitcher.devices import SwitcherV2Device
+
+    def __init__(self, hass: HomeAssistant, name: str, icon: Optional[str],
```

Comment:
I think `self.hass` is not yet created when inside `__init__`.
I needed hass to do a `hass.bus.async_listen`.
Although, after your previous change requests, this might not be needed anymore.

Nevertheless, I think if there is a need to use `hass` inside the `__init__` function, it's more elegant to pass it a argument.
But I can easily change the design to avoid that if that's a problem.

Reply:
Fixed in this [commit](https://github.com/home-assistant/home-assistant/pull/22325/commits/6dc98e87e893f7a75003f4f82da546824b834723).